hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/SevenToEightUpgradeHandler.java

```diff
@@ -62,19 +78,38 @@ public class SevenToEightUpgradeHandler implements UpgradeHandler {
   @Override
   public Map<ConfigProperty, String> upgrade(HoodieWriteConfig config, HoodieEngineContext context,
                                              String instantTime, SupportsUpgradeDowngrade upgradeDowngradeHelper) {
+    Map<ConfigProperty, String> tablePropsToAdd = new HashMap<>();
     HoodieTable table = upgradeDowngradeHelper.getTable(config, context);
     HoodieTableMetaClient metaClient = table.getMetaClient();
     HoodieTableConfig tableConfig = metaClient.getTableConfig();
+    // If auto upgrade is disabled, set initial version and writer version to 6 and return
+    if (!config.autoUpgrade()) {
+      setInitialVersion(config, table.getMetaClient().getTableConfig(), tablePropsToAdd);
+      config.setValue(HoodieWriteConfig.WRITE_TABLE_VERSION, String.valueOf(HoodieTableVersion.SIX.versionCode()));
```

Comment:
So, the expectation is that subsequent operations after upgrade needs to use the same WriteConfig ? 
By default autoUpgrade is on and it will be turned off when user doesn't want upgrade to run. In this case, why do we would the case arise ? 

Wouldn't this step (https://github.com/apache/hudi/blob/aa0f57c60361a7c03ec99da56c2264452f89927d/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngrade.java#L64 ) avoid this case here ?

Reply:
That condition in `UpgradeDowngrade.needsUpgradeOrDowngrade` blocks the pipeline and does not even let the binary to be upgraded. This change is to facilitate migration protocol. As per the migration protocol, expectation is users can set autoUpgrade config to false and upgrade the binary, to continue writing in tv=6, then when all readers are upgrade, they can remove the autoConfig override.

Recommended migration steps:
1. Stop any async table services in 0.x completely. 
2. Upgrade writers to 1.x with tv=6 (this won't auto-upgrade anything); 0.x readers will continue to work; writers can also be readers and will continue to read both tv=6.
3. Upgrade table services to 1.x with tv=6, and resume operations. 
4. Upgrade all remaining readers to 1.x, with tv=6. 
5. Redeploy writers with tv=8; table services and readers will adapt/pick up tv=8 on the fly.

Reply:
Open questions:

1. If the table version is < 6 and user deploys 1.x code with autoUpgrade = false, should we upgrade to table version 6 and stop or error out asking user to upgrade with 0.15 code first ? 
2. Should we nuke metadata table when autoUpgrade == true ?
3. If we are supporting metadata table with writer version 6, we need to test this scenario or should we nuke metadata table  when 1.x code is deployed ?

Reply:
1. If autoUpgrade is false, pre-table version 6, throw error to turn autoupgrade and table version=6, to force an upgrade to 0.14/tv=6 first. And document that users need to upgrade to 0.14/0.15 first.
2. When tv = 6 with MT, autoUpgrade is true, then upgrade to tv = 8  including MT. Assuming all readers are upgraded, this should work.
3. When tv = 6, with MT, autoUpgrade is false, then do nothing, just simple package upgrade and continue to write in tv = 6 (without MT). And add a check in upgrade path to later blow away MT if behind DT. Operating like this in 3, will eventually diverge MT/DT. So during the actual upgrade, when users remove the autoUpgrade override, MT needs to be rebuilt).

I have update the `UpgradeDowngrade.needsUpgradeOrDowngrade` and `SevenToEightUpgradeHandler` accordingly.